mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/mail/PostgresMessageMapper.java

```diff
@@ -0,0 +1,373 @@
+                return message;
+            })
+            .flatMap(mailboxMessage -> setNewMessageUid(mailboxMessage)
+                .then(setNewMessageModSeq(mailboxMessage)))
+            .then(mailboxMessageDAO.insert(message))
+            .then(saveBodyContent(message)
+                .flatMap(blobId -> messageDAO.insert(message, blobId.asString())))
+            .thenReturn(message.metaData());
+    }
+
+    private Mono<BlobId> saveBodyContent(MailboxMessage message) {
+        return Mono.fromCallable(() -> MESSAGE_CONTENT_LOADER.apply(message))
+            .flatMap(bodyByteSource -> Mono.from(blobStore.save(blobStore.getDefaultBucketName(), bodyByteSource, LOW_COST)));
+    }
+
+    @Override
+    public Iterator<UpdatedFlags> updateFlags(Mailbox mailbox, FlagsUpdateCalculator flagsUpdateCalculator, MessageRange range) {
+        return updateFlagsPublisher(mailbox, flagsUpdateCalculator, range)
+            .toIterable()
+            .iterator();
+    }
+
+    @Override
+    public Mono<List<UpdatedFlags>> updateFlagsReactive(Mailbox mailbox, FlagsUpdateCalculator flagsUpdateCalculator, MessageRange range) {
+        return updateFlagsPublisher(mailbox, flagsUpdateCalculator, range)
+            .collectList();
+    }
+
+    private Flux<UpdatedFlags> updateFlagsPublisher(Mailbox mailbox, FlagsUpdateCalculator flagsUpdateCalculator, MessageRange range) {
+        return mailboxMessageDAO.findMessagesMetadata((PostgresMailboxId) mailbox.getMailboxId(), range)
+            .flatMap(currentMetaData -> updateFlags(currentMetaData, flagsUpdateCalculator));
+    }
+
+    private Mono<UpdatedFlags> updateFlags(ComposedMessageIdWithMetaData currentMetaData, FlagsUpdateCalculator flagsUpdateCalculator) {
+        Flags oldFlags = currentMetaData.getFlags();
+        Flags newFlags = flagsUpdateCalculator.buildNewFlags(oldFlags);
+
+        if (oldFlags.equals(newFlags)) {
+            return Mono.empty();
+        }
+
+        ComposedMessageId composedMessageId = currentMetaData.getComposedMessageId();
+
+        UpdatedFlags updatedFlags = UpdatedFlags.builder()
+            .messageId(composedMessageId.getMessageId())
+            .modSeq(currentMetaData.getModSeq())
+            .oldFlags(oldFlags)
+            .newFlags(newFlags)
+            .uid(composedMessageId.getUid())
+            .build();
+
+        return mailboxMessageDAO.updateFlag((PostgresMailboxId) composedMessageId.getMailboxId(), composedMessageId.getUid(), updatedFlags)
+            .then(Mono.just(updatedFlags));
+
+    }
+
+    @Override
+    public Mono<List<UpdatedFlags>> resetRecentReactive(Mailbox mailbox) {
+        return mailboxMessageDAO.findAllRecentMessageMetadata((PostgresMailboxId) mailbox.getMailboxId())
+            .flatMap(mailboxMessage -> updateFlags(mailboxMessage,
+                new FlagsUpdateCalculator(new Flags(Flags.Flag.RECENT), MessageManager.FlagsUpdateMode.REMOVE)))
+            .collectList();
```

Comment:
```suggestion
        return mailboxMessageDAO.findAllRecentMessageMetadata((PostgresMailboxId) mailbox.getMailboxId())
            .collectList()
            .flatMap(mailboxMessages -> updateFlags(mailboxMessages,
                new FlagsUpdateCalculator(new Flags(Flags.Flag.RECENT), MessageManager.FlagsUpdateMode.REMOVE)))
            .collectList();
```

We want to update all messages at once... So that we generate a single MODSEQ.